plugins/rebalance/rest.go

```diff
@@ -0,0 +1,253 @@
+package rebalance
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gluster/glusterd2/glusterd2/gdctx"
+	restutils "github.com/gluster/glusterd2/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/pkg/api"
+	rebalanceapi "github.com/gluster/glusterd2/plugins/rebalance/api"
+	"github.com/gorilla/mux"
+	"github.com/pborman/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+func createRebalanceInfo(rebalance *rebalanceapi.RebalanceInfo) (*rebalanceapi.RebalanceInfo, error) {
+	v := new(rebalanceapi.RebalanceInfo)
+	v.Volname = rebalance.Volname
+	v.RebalanceID = uuid.NewRandom()
+	v.Status = rebalance.Status
+	v.Cmd = rebalance.Cmd
+	v.CommitHash = rebalance.CommitHash
+	return v, nil
+}
+
+func rebalanceStart(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	// collect inputs from url
+	volname := mux.Vars(r)["volname"]
+
+	var req rebalanceapi.RebalanceStartReq
+
+	// Unmarshal Request so to handle fix-layout and start force
+	if err := restutils.UnmarshalRequest(r, &req); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusUnprocessableEntity, "Unable to parse the request", api.ErrCodeDefault)
+		return
+	}
+
+	vol, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(r.Context(), w, http.StatusNotFound, "Invalid volume", api.ErrCodeDefault)
+		return
+	}
+
+	if vol.State != volume.VolStarted {
+		restutils.SendHTTPError(r.Context(), w, http.StatusBadRequest, "Volume not started", api.ErrCodeDefault)
+		return
+	}
+
+	if vol.DistCount == 1 {
+		restutils.SendHTTPError(r.Context(), w, http.StatusBadRequest, "Volume is not distributed volume or contain only 1 brick",
+			api.ErrCodeDefault)
+		return
+	}
+
+	// Check for remove- brick pending
+	//TODO
+
+	//A simple transaction to start rebalance
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+	lock, unlock, err := transaction.CreateLockSteps(volname)
+	if err != nil {
+		restutils.SendHTTPError(r.Context(), w, http.StatusInternalServerError, err.Error(), api.ErrCodeDefault)
+		return
+	}
+
+	txn.Nodes = vol.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc: "rebal-start.Commit",
+			Nodes:  txn.Nodes,
+		},
+		{
+			DoFunc: "rebal-start.StoreVolume",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
+		},
+		unlock,
+	}
+
+	err = txn.Ctx.Set("volname", volname)
+	if err != nil {
+		logger.WithError(err).Error("failed to set request in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err.Error(), api.ErrCodeDefault)
+		return
+	}
+	rebalance := new(rebalanceapi.RebalanceInfo)
+	rebalance.Volname = volname
+	rebalance.Status = rebalanceapi.GfDefragStatusStarted
+	rebalance.Cmd = rebalanceapi.GfDefragCmdStart
+	setCommitHash(rebalance)
+	rebal, err := createRebalanceInfo(rebalance)
```

Comment:
How about this? instead of multiple instance of reblanceinfo struct

    func createRebalanceInfo(volname string) *rebalanceapi.RebalanceInfo {
        return &rebalanceapi.RebalanceInfo {
     	    Volname: volname,
     	    RebalanceID: uuid.NewRandom()
     	    Status: rebalanceapi.GfDefragStatusStarted,
     	    Cmd: rebalanceapi.GfDefragCmdStart,
     	    CommitHash: getCommitHash(),
        }
    }

Reply:
that works fine. I made necessary changes